pkg/image/controller/signature/signature_import_controller.go

```diff
@@ -0,0 +1,185 @@
+		ctx:            ctx,
+		fetchTimeout:   fetchTimeout,
+		imageClient:    imageClient,
+		imageCache:     imageInformer.Informer().GetStore(),
+		imageHasSynced: imageInformer.Informer().HasSynced,
+	}
+
+	imageInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			image := obj.(*imageapi.Image)
+			glog.V(4).Infof("Adding image %s", image.Name)
+			controller.enqueueImage(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			image := cur.(*imageapi.Image)
+			glog.V(4).Infof("Updating image %s", image.Name)
+			controller.enqueueImage(cur)
+		},
+	}, resyncInterval)
+
+	return controller
+}
+
+func (s *SignatureImportController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer s.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(stopCh, s.imageHasSynced) {
+		return
+	}
+
+	glog.V(5).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+
+}
+
+func (s *SignatureImportController) worker() {
+	for {
+		if !s.work() {
+			return
+		}
+	}
+}
+
+// work returns true if the worker thread should continue
+func (s *SignatureImportController) work() bool {
+	key, quit := s.queue.Get()
+	if quit {
+		return false
+	}
+	defer s.queue.Done(key)
+
+	if err := s.syncImageSignatures(key.(string)); err == nil {
+		// this means the request was successfully handled.  We should "forget" the item so that any retry
+		// later on is reset
+		s.queue.Forget(key)
+	} else {
+		// if we had an error it means that we didn't handle it, which means that we want to requeue the work
+		utilruntime.HandleError(fmt.Errorf("error syncing service, it will be retried: %v", err))
+		s.queue.AddRateLimited(key)
+	}
+	return true
+}
+
+func (s *SignatureImportController) enqueueImage(obj interface{}) {
+	_, ok := obj.(*imageapi.Image)
+	if !ok {
+		return
+	}
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	s.queue.Add(key)
+}
+
+func (s *SignatureImportController) syncImageSignatures(key string) error {
+	obj, exists, err := s.imageCache.GetByKey(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve image %v from store: %v", key, err)
+		return err
+	}
+	if !exists {
+		glog.V(4).Infof("Image has been deleted %v", key)
+		return nil
+	}
+
+	image := obj.(*imageapi.Image)
+	signatures, err := s.fetchImageSignatures(image)
+	if err != nil {
+		glog.V(4).Infof("Failed to fetch image %s signatures: %v", image.Name, err)
+		return err
+	}
+	if len(signatures) == 0 {
+		glog.V(5).Infof("Image %s has not signatures to import", image.Name)
+		return nil
+	}
+
+	t, err := kapi.Scheme.DeepCopy(image)
+	if err != nil {
+		return err
+	}
+
+	updatedImage := t.(*imageapi.Image)
+	updatedImage.Signatures = signatures
```

Comment:
is this overwriting image signatures that a user manually added?

Reply:
@mtrmac I want to sort out how we are going to sync the signatures in this controller.

We can add a common prefix for automatically imported signatures which we can safely remove when the sync runs and replace with new signatures (or delete them when we don't receive any from sigstore).  That will make sure we won't touch signatures created by users manually.

Reply:
Only adding signatures, never removing them, would be a good 90% solution I can be very happy with.

The remaining 10/90% is marking a signature as obsolete, and probably not serving it to clients, when it is removed from the original registry (think a signature for `:latest`, which should be served only for the truly latest image). This is more complex if several repos serve the same image (manifest digest) and the same signature, and only one of them stops serving that signature; it’s not really obvious what to do in that case. Ideally, I guess, the signature status for each imagestream should mirror the signature status of the upstream repository for _that_ imagestream, but that seems like a pretty big change to the model.

Reply:
> We can add a common prefix for automatically imported signatures

I’m not entirely sure what you mean; is that a constant like `imported-signature`, or some identification of the specific image stream?